core/src/main/java/org/apache/calcite/plan/volcano/CascadeRelSubset.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.plan.volcano;
+
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptCost;
+import org.apache.calcite.plan.RelTraitSet;
+
+/***
+ * An implementation of RelSubset for CascadePlanner.
+ * Holding the optimizing state.
+ *
+ * In cascade, a subset is a group with required physical properties (TraitSet)
+ */
+public class CascadeRelSubset extends RelSubset {
+
+  /**
+   * An enum representing optimizing state of current RelSubset
+   */
```

Comment:
What's the difference between `CascadeRelSubset` and `RelSubset`?

Reply:
CascadeRelSubset have two more fields:  optimizeState and upperBound.  The comments in this class will help you understand what they are used for.